Use parking_lot::RwLock for PropertyDeclarationBlock

SimonSapin opened the pull request:
<!-- Please describe your changes on the following line: -->

As discussed in https://bugzilla.mozilla.org/show_bug.cgi?id=1305141

Closes #13176

---

Original PR title: Stop relying on `impl<T: HeapSizeOf> HeapSizeOf for Arc<T>`
https://github.com/servo/heapsize/issues/37#issuecomment-249861171

This builds on top of that.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [x] These changes do not require tests because refactor

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13459)

<!-- Reviewable:end -->

SimonSapin committed:
Use PropertyDeclarationBlock in a DOMRefCell everywhere.

SimonSapin committed:
Make style build without `impl<T: HeapSizeOf> HeapSizeOf for Arc<T>`.

The removal of this impl is not included in this commit.
CC https://github.com/servo/heapsize/issues/37#issuecomment-249861171

SimonSapin committed:
Make script build without `impl<T: HeapSizeOf> HeapSizeOf for Arc<T>`.

The removal of this impl is not included in this commit.
CC https://github.com/servo/heapsize/issues/37#issuecomment-249861171

SimonSapin committed:
Update heapsize

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/Cargo.toml, components/style/restyle_hints.rs, components/style/properties/properties.mako.rs, components/style/selector_matching.rs, components/style/stylesheets.rs, components/style/keyframes.rs
 * @kichjang: components/script/dom/document.rs, components/script/dom/htmllinkelement.rs, components/script/dom/element.rs, components/script/dom/htmlstyleelement.rs, components/script/Cargo.toml, components/script/dom/workerglobalscope.rs, components/script/dom/htmlmetaelement.rs, components/script/dom/htmlimageelement.rs, components/script_layout_interface/Cargo.toml, components/script/dom/htmlobjectelement.rs

nox commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit a710b40 has been approved by `nox`

<!-- @bors-servo r=nox a710b4031fa9c456c0a91b26f3bec0248b41b0a6 -->

SimonSapin committed:
Use parking_lot::RwLock instead of DOMRefCell for PropertyDeclarationBlock

SimonSapin committed:
Move DOMRefCell back into script.

We’re not using it in style after all.

SimonSapin committed:
Revert "Make DOMRefCell use style’s copy of RefCell"

This reverts commit ec723057b2360446790d436fe5e234d798fb4175.

nox commented:
Reviewed 3 of 3 files at r1, 12 of 12 files at r2, 8 of 8 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13459)*
<!-- Sent from Reviewable.io -->

SimonSapin committed:
Don’t pretend DOMRefCell is Send or Sync, after all.

emilio commented on components/script/dom/cssstyledeclaration.rs:
```diff
@@ -119,43 +118,42 @@ impl CSSStyleDeclarationMethods for CSSStyleDeclaration {
         // Step 2
         if let Some(shorthand) = Shorthand::from_name(&property) {
+            let style_attribute = owner.style_attribute().borrow();
+            let style_attribute = if let Some(ref style_attribute) = *style_attribute {
```
This might be easier to read with a `match` instead?

emilio reviewed (approved):
I'm fine with this. In general, I'd like us to ensure we don't block in most cases (it should only be needed for keyframes), but I see why that's hard. It can become a followup. In any case it's better than the YOLO RefCell we have now.

SimonSapin commented:
@bors-servo r=emilio p=10

(Prioritized as this is prone to bitrot.)

bors-servo commented:
:pushpin: Commit d6538f9 has been approved by `emilio`

<!-- @bors-servo r=emilio d6538f9d8fdd118985ca1308633cf34ce55af8b1 -->

bors-servo commented:
:hourglass: Testing commit d6538f9 with merge 0e870bf...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3274)

SimonSapin commented:
@bors-servo r=emilio

Fixed order of `use` statements.

bors-servo commented:
:pushpin: Commit 90d245a has been approved by `emilio`

<!-- @bors-servo r=emilio 90d245a881f9d96e472df8cc0fdc20b29c9e36a2 -->

bors-servo commented:
:hourglass: Testing commit 90d245a with merge 4c12a49...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3457)

SimonSapin commented:
Uh oh.

```
  ▶ FAIL [expected PASS] /_mozilla/css/iframe/hide_after_load.html
  └   → /_mozilla/css/iframe/hide_after_load.html 7e556c519a46ddc020581d2275b67ff4c39cd5d8
/_mozilla/css/iframe/hide_after_load_ref.html ce129a418aad4e23134dc79329a8508de6be8223
Testing 7e556c519a46ddc020581d2275b67ff4c39cd5d8 == ce129a418aad4e23134dc79329a8508de6be8223

  ▶ Unexpected subtest result in /_mozilla/mozilla/calc.html:
  │ FAIL [expected PASS] calc(10px - 5px)
  │   → assert_equals: expected "5px" but got "20px"
  │ FAIL [expected PASS] calc(10px / 2)
  │   → assert_equals: expected "5px" but got "20px"
  │ FAIL [expected PASS] calc(10px + 10em - 10px)
  │   → assert_equals: expected "160px" but got "10px"
  │ FAIL [expected PASS] calc(1px + 1pt + 1pc + 1in + 1cm + 1mm)
  │   → assert_equals: expected "155.88333333333333px" but got "20px"
  │ FAIL [expected PASS] calc(((3 - 1) * (8 + 4)) * 10px)
  │   → assert_equals: expected "240px" but got "170px"
  │ FAIL [expected PASS] calc(5 * (20px / 2 + 7 * (3em + 12px/4 + (8 - 2) * 2rem)))
  │   → assert_equals: expected "8555px" but got "10px"
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:45:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:42:5
  └ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:41:1

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-clip-001.htm:
  │ FAIL [expected PASS] background-clip_content-box
  │   → assert_equals: background-clip supporting value expected "content-box" but got "border-box"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-clip-001.htm:21:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-clip-001.htm:20:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-clip-001.htm:
  │ FAIL [expected PASS] background-clip_padding-box
  │   → assert_equals: background-clip supporting value expected "padding-box" but got "border-box"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-clip-001.htm:15:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-clip-001.htm:14:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-origin-001.htm:
  │ FAIL [expected PASS] background-origin_padding-box
  │   → assert_equals: background-origin supporting value expected "padding-box" but got "border-box"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-origin-001.htm:15:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-origin-001.htm:14:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-size-001.htm:
  │ FAIL [expected PASS] background-size_auto_auto
  │   → assert_equals: background-size supporting value expected "auto auto" but got "contain"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:69:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:68:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-size-001.htm:
  │ FAIL [expected PASS] background-size_length_percentage
  │   → assert_equals: background-size supporting value expected "15px 50%" but got "0px auto"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:99:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:98:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-size-001.htm:
  │ FAIL [expected PASS] background-size_percentage_auto
  │   → assert_equals: background-size supporting value expected "50% auto" but got "0px auto"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:105:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:104:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-size-001.htm:
  │ FAIL [expected PASS] background-size_auto_percentage
  │   → assert_equals: background-size supporting value expected "auto 50%" but got "0px auto"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:81:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:80:9
```

I don’t know what’s going on here, but there’s no known intermittent filed for these. Will need to investigate.

SimonSapin commented:
The only (I think) behavior change is that this PR removes some `Arc::make_mut` and uses locking instead. So a `PropertyDeclarationBlock` that could be referenced elsewhere and was previously immutable is now mutable. This seemed like a good idea, reducing the amount of allocation and cloning.

Disabling the applicable declaration cache (by making `PartialEq for ApplicableDeclarationsCacheQuery` always return false) seems to fix the issue. This is likely because the cache uses the address of a `Arc<PropertyDeclarationBlock>` (now `Arc<RwLock<PropertyDeclarationBlock>>`) pointer as its hash / identity.

I can’t think of proper fix that doesn’t revert to the previous copy-on-write behavior of using `Arc::make_mut` (or equivalent) in CSSOM (including `Element.style`), cloning blocks when modifying them if they have more than one reference.

@pcwalton, any idea?

emilio commented:
@SimonSapin: Flushing the applicable declarations cache after a restyle should be fine, right?

SimonSapin commented:
Interesting. Or maybe have any `PropertyDeclarationBlock` mutation mark the cache as dirty? That way we can still keep it between restyles for DOM updates.

emilio commented:
Por even better, maybe just avoid adding to the applicable declaration cache if the matched element has a style attribute?

Seems like it's impossible for that to match in the same restyle with any other element, and I don't think we care a lot about the case where there are such a little amount of elements that the entry persists in the cache across restyles, so it might even be a win?

SimonSapin commented:
We’ll still need some way to deal with this cache when style rules are mutated, when we eventually support CSSOM.

SimonSapin committed:
Disable applicable declarations cache for elements with a style attribute.

The cache is never effective in this case since a style attribute
adds to the set of applicable declarations.

This also avoid avoids the cache giving out-dated (incorrect) results
when `PropertyDeclarationBlock` becomes mutable in a later commit.
This is sufficient for now since CSSOM is only implemented for
`Element.style`. When StyleRule mutation is added,
some additional mechanism (such as evicting the cache) will be needed.

SimonSapin commented:
Rebased and added a "Disable applicable declarations cache for elements with a style attribute" commit.

We’ll need to do something (maybe mark the cache as dirty whenever a `PropertyDeclarationBlock` is mutated?) when more of CSSOM is implemented, but in the meantime I’d like to land this before it bit-rots again.

re-r? @emilio

emilio commented:
Also, I think this is the only not-previously-reviewed commit, so feel free to r=me the whole PR when that comment is added.

@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 3d1a917 with merge fc84d1e...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3310), [arm64](http://build.servo.org/builders/arm64/builds/3323), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3400), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/17), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/17), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3387), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3367), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/68), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/67), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1275)

SimonSapin commented:
I, that was the only new significant change.

@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit d15ac9b has been approved by `emilio`

<!-- @bors-servo r=emilio d15ac9be55027b901eeaa2a36ec7976caf18b4cc -->

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #13570

bors-servo commented:
:pushpin: Commit d15ac9b has been approved by `emilio`

<!-- @bors-servo r=emilio d15ac9be55027b901eeaa2a36ec7976caf18b4cc -->

bors-servo commented:
:hourglass: Testing commit d15ac9b with merge aea9545...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3313), [arm64](http://build.servo.org/builders/arm64/builds/3326), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3403), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/20), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/20), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3390), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3370), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/71), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/70), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1278)
